pkg/log/logger.go

```diff
@@ -0,0 +1,224 @@
+package log
+
+import (
+	"context"
+	"fmt"
+	"github.com/applike/gosoline/pkg/clock"
+	"os"
+	"strings"
+)
+
+const (
+	LevelTrace = "trace"
+	LevelDebug = "debug"
+	LevelInfo  = "info"
+	LevelWarn  = "warn"
+	LevelError = "error"
+
+	PriorityTrace = 0
+	PriorityDebug = 1
+	PriorityInfo  = 2
+	PriorityWarn  = 3
+	PriorityError = 4
+)
+
+var levelNames = map[int]string{
+	PriorityTrace: LevelTrace,
+	PriorityDebug: LevelDebug,
+	PriorityInfo:  LevelInfo,
+	PriorityWarn:  LevelWarn,
+	PriorityError: LevelError,
+}
+
+var levelPriorities = map[string]int{
+	LevelTrace: PriorityTrace,
+	LevelDebug: PriorityDebug,
+	LevelInfo:  PriorityInfo,
+	LevelWarn:  PriorityWarn,
+	LevelError: PriorityError,
+}
+
+func LevelName(level int) string {
+	return levelNames[level]
+}
+
+func LevelPriority(level string) int {
+	return levelPriorities[level]
+}
+
+type Data struct {
+	Context       context.Context
+	Channel       string
+	ContextFields map[string]interface{}
+	Fields        map[string]interface{}
+}
+
+type Fields map[string]interface{}
+
+//go:generate mockery --name Logger
+type Logger interface {
+	Debug(format string, args ...interface{})
+	Info(format string, args ...interface{})
+	Warn(format string, args ...interface{})
+	Error(format string, args ...interface{})
+
+	WithChannel(channel string) Logger
+	WithContext(ctx context.Context) Logger
+	WithFields(Fields) Logger
+}
+
+type GosoLogger interface {
+	Logger
+	Option(opt ...Option) error
+}
+
+type LoggerSettings struct {
+	Handlers []HandlerSettings `cfg:"handlers"`
+}
+
+type HandlerSettings struct {
+	Type string `cfg:"type"`
+}
+
+type gosoLogger struct {
+	clock        clock.Clock
+	data         Data
+	ctxResolvers []ContextFieldsResolver
+	handlers     []Handler
+}
+
+func NewLogger() *gosoLogger {
+	return NewLoggerWithInterfaces(clock.NewRealClock(), []Handler{})
+}
+
+func NewLoggerWithInterfaces(clock clock.Clock, handlers []Handler) *gosoLogger {
+	return &gosoLogger{
+		clock: clock,
+		data: Data{
+			Context:       nil,
+			Channel:       "main",
+			ContextFields: make(map[string]interface{}),
+			Fields:        make(map[string]interface{}),
+		},
+		ctxResolvers: nil,
+		handlers:     handlers,
+	}
+}
+
+func (l *gosoLogger) Option(options ...Option) error {
+	for _, opt := range options {
+		if err := opt(l); err != nil {
+			return fmt.Errorf("can not apply option %T", err)
```

Comment:
Why %T? Shouldn't it be %w to embed the error?

Reply:
fixed